### ISSUE TITLE
CommandTimeout parameter and readme issues

### ISSUE THREAD

**Issue by cmp-202** [cmp-202/ssh2shell#27]
Invalid Issue Path
----------------
The definition of this issues has been altered due to changing function parameters breaking existing scripts using ssh2shell.

Instead a better method for using commandTimeout has been provided in test/timeouttest.js and updated info in the readme. 

* sshObj can be accessed by using the instance event handler along with event emiters and the this keyword. None of those can be accessed from the host.oncommandTimeout handler.
* It is not possible using the host.onCommandTimeout event to retrigger the timeout timer without getting a response from the host after using the stream.write() command. 
* To close the connection when using the host.onCommandTimeout event call both stream.end() and connection.close(). The stream.end() call will handle returning the sessionText and connection.close() will terminate the connection.
* Alterations to branch were returned to match master except for additions and corrections that didn't break the existing function definition.
* Branch has been merged into master.
 
The following will not be implemented
-------------
The event handler onCommandTImeout has parameters that need to be changed to give access to the commands array in the sshObj. The readme also doesn't match the function definition in that the definition doesn't have sshObj as a parameter. The connection parameter is redundant as the connection can be closed using stream.end().
New branch created to produce changes

**Comment by ajiratech2** [comment]
I tried it and the new branch fails with error:

"   if (_this.sshObj.keyboard - interactive) {
ReferenceError: interactive is not defined
    at SSH2Shell.<anonymous> (C:\Users\ahujapa\Documents\workarea\Automation tools\Projects\ConsoleServer\Node\node_modules\ssh2shell-CommandTimeout\lib\ssh2shell.js:309:39)
    at emitMany (events.js:108:13)
    at SSH2Shell.emit (events.js:182:7)"

So I figured that the "Keyboard-interactive" issue must not have merged in master copy that was used to create a branch for this. I made changes to fix it on my end to continue testing and ran into further issues. 
Line 309 and 310 updated to:
          if (_this.sshObj["keyboard-interactive"]) {
            return _this.sshObj["keyboard-interactive"](name, instructions, instructionsLang, prompts, finish);
          }

Now I move past and get a different error:

"message: Connected
message: Connection error: Error: Timed out while waiting for handshake
message: Closed"

I'm not sure why I'm getting this handshake error. I've compared the code for this branch with the "connection-options" and do not see any other differences (except for 'onCommandTimeout', which is intended).

**Comment by ajiratech2** [comment]
I did some more troubleshooting and it seems as though it is timing out while waiting for Handshake because the "keyboard-interactive" event is not implemented properly. The changes I made to the downloaded branch must be missing something.

so I copied the code for "onCommandTimeout" from the new branch over to the old ssh2shell-ssh2.connect-options and now I get the following error:

TypeError: stream.write is not a function

I'll continue to play with it and report back.

**Comment by cmp-202** [comment]
Hmm not seeing any issues in the coffeescript and the only difference I see between the two branches is the changes to onCommandTimeout with those parameter differences.

**Comment by cmp-202** [comment]
Can you send me your script.

**Comment by ajiratech2** [comment]
Sure. My script at the end of this comment.

My last test suggests that in onCommandTimeout event there is no stream.write(). I get the following error for it:

"TypeError: stream.write is not a function"




---------  script begins ----------

    var connectViaSSH = function(endHost,port,arrayOfCommands,callback) {

        var listOfCommands = arrayOfCommands.slice(0,arrayOfCommands.length);       
                            // creating a clone because the commands are removed from passed variable as they are executed.

        var host = {
            server: {
                host: endHost,
                port: port,
                userName: username,
                password: password,
                hashMethod:     "md5", 
                readyTimeout: readyTimeout,
                tryKeyboard: true,
                algorithms: {
                    kex: arrayOfKex,
                    cipher: arrayOfCipher
                    }
                },
                standardPrompt: customStandardPrompt,
                commands: listOfCommands,
                msg: {
                    send: function( message ) {
                    console.log("message: " + message);
                    }
                },
                verbose: verboseStatus,
                debug: debugStatus,
                idleTimeOut: idleTimeout,
                ["keyboard-interactive"]: function(name, instructions, instructionsLang, prompts, finish){
                    console.log('Connection :: keyboard-interactive');
                    finish([password]);
                },
                onCommandComplete:   function( command, response, sshObj ) {
                    console.log("-------> completed command : " + command);

                }, 
                onCommandTimeout: function( command, response, stream, sshObj ) {

                   if (command === "" && response === "Connected to port 22." ) {

                     console.log(" yes it matched, sending newline.");

                     stream.write('\n'); //to trigger the standard prompt on a new line

                   }else{
                     console.log("It did not match, ending session.");
                     stream.end();
                   }
                }, 
                onEnd: function( sessionText, sshObj ) {
                    //sshObj.msg.send("--------- onEnd has ------------");
                    //sshObj.msg.send(sessionText);

                    callback(0,sessionText); // this sends the commands and output generated
                }

            };

        //Create a new instance
        var SSH2Shell = require ('ssh2shell-ssh2.connect-options'),
        // var SSH2Shell = require ('ssh2shell-commandtimeout'),
        SSH = new SSH2Shell(host);

        //Start the process
        SSH.connect();

    }
--- script ends -----



P.S> - When I do a compare of "CommandTimeout" branch with "Master" for coffeescript I see no differences now. ( not even for onCommandTimeout).

**Comment by cmp-202** [comment]
The params order in commandTimeout is not correct it should be function( command, response, sshObj, stream )

**Comment by ajiratech2** [comment]
Thanks. that cleared the error message I received for "stream.write" .
but nothing happens after that. It does not look like stream.write() pushed the command. I say this because after another period of idleTimeout I do not see the script entering "onCommandTimeout" again.

The script is just hung right now. Do we need to force a call to something like "_processNextCommand" ?

**Comment by cmp-202** [comment]
So you tried both stream.write(“\n”) or stream.write(“\n\n”) and stream.write(command) [ command from sshObj.commands.shift()]. I would also output console.log(response) so you can see if anything came back into the buffer.

I am just setting up my linux VM to use keyboard-interaction so I can test it directly, hopefully that will shed some light on it.

**Comment by ajiratech2** [comment]
I just tried stream.write("\n\n") no luck. When I try " command = sshObj.commands.shift(); " I get the following error:

command = sshObj.commands.shift();
                                                  ^
TypeError: Cannot read property 'shift' of undefined

Then I tried to send another command like this stream.write("show version\n"); it still seems like it is hung.

I am logging "command" and "response" in onCommandTimeout, but it does not come back in this event. If I let it sit then after several minutes I see IdleTimeout hit.

"message: Connection error: Error: Idle Timeout"

**Comment by ajiratech2** [comment]
the carrot for the error was supposed to point to start of "shift()".

My idle timeout is set to 10000 (i.e. 10 seconds).

I think stream.write() alone is not sending the command.

**Comment by ajiratech2** [comment]
I added event "onCommandProcessing" and I see that after the first "onCommandTimeout" it does not fire "onCommandProcessing" again.

**Comment by ajiratech2** [comment]
Is it possible that stream.write("\n") does not translate into a newline character ?

I ask this because I know if I hit any other key using my ssh terminal program then it does not proceed. It has to be a newline.

**Comment by cmp-202** [comment]
Cheers for that very helpful info

I have a workaround that might be worth trying:

Remove the CommandTimeout handler and replace it with

 

onCommandProcessing: function( command, response, sshObj, stream ){

   if ( response.indexOf(“Connected”) != -1 ){

      stream.write(“\n”);

   }

}

 

Leave the rest of it with me for now if that is ok. I need to get my server accepting keyboard-interactive so I can work out how to respond to this process. The timeout function is clearly not the place to handle it

**Comment by cmp-202** [comment]
So I finally got it going on my server and I think the onCommandProcessing method is the way to go for you.

When I successfully connect with keyboard-interactive the result is a standard prompt for me but it isn’t for you.

onCommandProcessing is the event handler that runs every time a char is added to the buffer once connection is made and any output from the host is received including banners and prompt. By detecting “Connected” or the full text “Connected on port 22” you can respond via stream.write() with whatever will bring up the standard prompt for the process to continue and avoid a timeout. Unlike the onCommandTimeout handler output using stream.write() will continue the loop process of checking for the prompt and respond by sending a command. Not sure why that didn’t happen with onCommandTimeout might look that.

 

I am unsure how you managed to use the host.keyboard-interactive definition because when I ran my test script it refused to allow the ‘-‘ so I have changed the code to use host.onKeyboardInteractive. Adding an instance event handler still works as SSH.on(“keyboard-interactive”, function (blar,blar){});. I will update the readme with keyboard-interactive details, the onCommandProcessing method if the result is not standard and correct the host event handler definition.

**Comment by cmp-202** [comment]
Spent quite a bit of time on this over the weekend. I have created/modified the [test/keyboard-interactivetest.j](https://github.com/cmp-202/ssh2shell/blob/master/test/keyboard-interactivetest.js) file to show the implementation of keyboard-interactive authentication and more complex command timeout event handling that can trigger a second timeout timer to stop the script hanging if there is no response from the server. Doesn't answer why `\n` or `\r\n` didn't work but if the server has a delay before it responds (given it is slow) then this method will help allow for that. By setting the timeout value to whatever and using the commandTimeout event to respond after that time any delay can be allowed for. I have tested this on my VM as best I could by changing standard prompt setting and what I was responding with stream.write() to generate passes, fails and command fails so all three worked for me in my context. Readme is updated, I have merged the changes into master but I have not published to npm yet.

**Comment by ajiratech2** [comment]
Thank you! I really appreciate your time and effort over the weekend.

I tried to pull it down and VPN to work so I could test it out, but my PC is shutdown and is sitting on the bios password prompt. (This happens every weekend, PC team applies patches and reboots automatically).

I'll have to try this first thing Monday morning and will report back.

Thanks,
Pankaj

**Comment by ajiratech2** [comment]
I pulled down the 'master' and tried it a few different ways and it still doesn't work. In your test when you do a stream.write() does that command actually becomes a part of "sessionText" and/or triggers another commandProcessing ?

In my case when I do stream.write("show version \n\n\n") in onCommandTimeout (or onCommandProcessing) it does not trigger another onCommandProcessing event. I'm looking into a way to log sessionText before and after stream.write() to see if the text/command we send through stream.write() makes it to the sessionText.

**Comment by ajiratech2** [comment]
I was able to do output of sshObj.sessionText in 'onCommandProcessing' and I do not see the command we send through stream.write() make it to sessionText neither in 'onCommandProcessing' 
nor in 'onCommandTimeout'

Here is how 'onCommandProcessing' looks like in my script. I'm doing the same again in 'onCommandTimeout'.

         onCommandProcessing:   function( command, response, sshObj, stream  ) {
               if (command === "" && response === "connected to port 22.") {
                    console.log(" yes it matched");
                    console.log("going to sleep for 5 seconds");
                    console.log("\n ------ session Text before sending command ---------\n" + sshObj.sessionText);
                    sleep(3000);
                    stream.write("show version\n\n\n");
                    stream.write("\n\n\n");
                    sleep(2000);
                    console.log("\n ------ session Text after sending command ---------\n" + sshObj.sessionText);

                }else{
                    console.log("It did not match, ending session.");
                    stream.end();
                }
       }

**Comment by ajiratech2** [comment]
Please let me know if there is any test you'd like me to do and share. I'm not sure what else I can try.

To me, it seems that stream.write() is not triggering another onCommandProcessing event.

I've run the test several times and watched the output very closely and it looks like that when the event 'onCommandProcessing' is triggered the first time, it is exactly at the time when the unusual prompt "Connected to port 22" is displayed. when using a commercial SSH client, it is at this moment that I hit "enter" and it then continues and gives a prompt for the actual device.

**Comment by ajiratech2** [comment]
Thank you for your help in trying to get this to work. I really appreciate the hours you put in to help me out.

Your wrapper of SSH2 is very good and very helpful in all other connections. This one situation is quite rare and it's alright if I can't get this to work.
I also tried Python Paramiko and got the same results, it fails at the same spot too.

I'll look for an alternative method to get this to work - maybe perl or bash has better luck. I'll be sure to post it on SO and here once I find it.

Thanks again!
Have a great weekend.

**Comment by ajiratech2** [comment]
Good News! I got it to work using your SSH2-shell wrapper itself.

On a hunch I decided to send 'onCommandProcessing' in the first dev branch you created for me 'ssh2shell-ssh2.connect-options' and there when I send newline, it goes through fine and does give a prompt as desired.

I repeated the same script by pointing it back to the 'ssh2-shell.master' and here it gets stuck again. So it looks like the original 'onCommandProcessing' is able to handle any flow adjustments just fine.

If the new master branch is handling onCommandProcessing differently than original branch, then it might be wise to go back to the original handling for your master branch.

Once again, thank you for your help in helping me move forward.

Please let me know if you'd like me to do any tests and share any results to ensure your master branch works well and continues to help a lot of people like myself.

**Comment by ajiratech2** [comment]
nevermind. It worked only a couple of times and now it is failing on the first branch as well.
I'll update if I get it to work reliably again.

**Comment by cmp-202** [comment]
Just had a couple of questions:

1.       What commercial ssh client where you using e.g. putty? 

2.       Do you connect via a serial com port to the hardware and then SSH into it? 

3.       If so is the com connection ssh2shell is communicating through set to the ideal com settings for the hardware(9600 baud rate and 8-n-1, flow control ). OR

4.       Is the settings in the commercial ssh program the same as your computers com connection settings? I know putty can be configured to handle the com connection settings itself. Might just be worth checking. 

 

If the commercial one works but everything else doesn’t then what is it about the commercial version connection that is different.

 

Re: \n end of line char.

Check out this link http://stackoverflow.com/questions/17393211/how-is-a-return-enter-keypress-encoded-in-a-string-sent-to-a-ssh-shell

It indicates that if you send ctrl-v and press enter that the shell will insert the next character literally instead of interpreting it as a command.

Might be worth connecting via your SSH client that works and see what results. 

When I do it in a ssh connection to Ubuntu server using putty it outputs ^m which is \r (“\x0d” hexdec) but in saying that I connect with ssh2shell using \n and it works. I have also read that “\r(null)” is another variance (“\x0d\x00”). 

 

I have also just added the ability to set host.enter = ”\n” so the end of line character can be changed to what every works for you.

Before it was hard coded because there had never been an issue. If it is different for you then you will need this setting.

 

Andrew

**Comment by cmp-202** [comment]
Timeout habdling

**Comment by cmp-202** [comment]
Timeout handling and scope/binding issues have been improved with access to `this` correctly referencing the instance and not sshObj (host object). The method for closing the connection has been improved. The test/timeouttest.js now provides a better example of how to use it in a number of ways. The problem of non-standard connection prompt "connected on port 22" that required an enter key response but wouldn't respond to stream.write("\n") was not resolved. Despite using different approaches we couldn't get the host to respond as required to run commands. The branch and its changes have been merged into master and this issue will be closed.

**Comment by ajiratech2** [comment]
I was able to get this to work correctly using your branch 'ssh2shell-ssh2.connect-options'.  The problem was a combination of two things - 
1. I needed to send '\r' in 'onCommandProcessing' after checking for 'Connected to 22.'
2. the newer branch for 'commandTimeout' and the 'master' are handling 'onCommandProcessing' differently.

On a hunch when I tried older branch 'ssh2shell-ssh2.connect-options' and sent '\r\n ' in 'onCommandProcessing' it started working.

Later I did some cleanup and changed too many things in the script and then it stopped working. It took me a few days to iron out other issues and then revisit this piece to get it to work.

Now it is working reliably when I use your older branch 'ssh2shell-ssh2.connect-options'  and send '\r' inside 'onCommandProcessing'. Here is the section of the script that is working reliably for me:

            onCommandProcessing:   function( command, response, sshObj, stream  ) {

                if (command === "" && response === "Connected to port 22.") {
                    stream.write("\r");
                    sshObj.msg.send("in 'onCommandProcessing' yes it matched. sending newline");
                }

            }, 
Thanks again for your time and effort in helping me get past these issues. This module now has all the features I need. Much appreciate it.

**Comment by cmp-202** [comment]
Just reviewed the difference between the two branch merges and yes the timeout branch was an utter botch job!!! Sorry, I reverted those parameter changes a few commits later. I would recommend giving master/npm ssh2shell a try there are some good improvements to closing connections, this keyword correctly referencing the instance in all host object function definitions giving easy access to this.emit() and other api functions and variables. Timeout fits the existing param content and order And I can’t see any reason for your code in commandProcessing to not work.

 

Andrew

**Comment by ajiratech2** [comment]
Sure. I'll give it a shot first thing in the morning.

Pankaj


On Sep 6, 2016, at 6:46 PM, Andrew McMurtrie <notifications@github.com<mailto:notifications@github.com>> wrote:

Just reviewed the difference between the two branch merges and yes the timeout branch was an utter botch job!!! Sorry, I reverted those parameter changes a few commits later. I would recommend giving master/npm ssh2shell a try there are some good improvements to closing connections, this keyword correctly referencing the instance in all host object function definitions giving easy access to this.emit() and other api functions and variables. Timeout fits the existing param content and order And I can't see any reason for your code in commandProcessing to not work.



Andrew

**Comment by ajiratech2** [comment]
Sorry, it took me a while to try out master and npm both. One of our important global offices went down due to MPLS circuit issues and I've had to spend all day today coming up with a contingency on how to bring it back online in situation like these.

When I try master or do an npm install, They're both failing for me with error :
"message: Error: Timed out while waiting for handshake, Level: client-timeout"
My ready-timeout is 45 seconds and I increased idle-timeout from 10 to 30 seconds , yet it still fails.

This error happens when I first connect to Lantronix - this part does not even require the special handling that I had to do in "onCommandProcessing". 

Here is a list of messages that are displayed when I use master:

message: Connected
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.onKeyboardInteractive
message: mdc-lab-slc-a.lab.wellmanage.com: this.onKeyboardInteractive
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.onError
message: mdc-lab-slc-a.lab.wellmanage.com: this.onError
message: Error: Timed out while waiting for handshake, Level: client-timeout
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.onClose
message: mdc-lab-slc-a.lab.wellmanage.com: this.onClose
message: Closed

I'll spend some more time on this to see if I can provide any more helpful pointers on where the issue might be.

**Comment by cmp-202** [comment]
[This is the updated test file](https://github.com/cmp-202/ssh2shell/blob/master/test/keyboard-interactivetest.js) with debug and verbose messages for the lastest version so you can check how I have done it.

**Comment by ajiratech2** [comment]
For some reason, this still fails in my environment. I enabled debugs and verbose on this script and did a test against the Lantronix device. Here are the messages I get:

C:\node\node_modules\ssh2shell\test> node keyboard-interactivetest.js

Connected
mdc-lab-slc-a.lab.wellmanage.com: Connection.onError
mdc-lab-slc-a.lab.wellmanage.com: this.onError
Error: Timed out while waiting for handshake, Level: client-timeout
mdc-lab-slc-a.lab.wellmanage.com: Connection.onClose
mdc-lab-slc-a.lab.wellmanage.com: this.onClose
Closed

You probably have better method of doing a compare between files. I did a compare using notepad++ between ssh2shell.coffee files of the two branches "master" vs "ssh2shell-ssh2.connect-options" and I see too many differences to list. 

Attaching the "ssh2shell-ssh2.connect-options" branch that you'd previously uploaded so you can do a comparison.

I think the code added from"commandTimedout" branch is not needed. The 'onCommandProcessing' event seems capable of special handling my environment and possibly other special case environments. When I first tried it before reporting this 'commandTimedout' issue I tried 'onCommandProcessing' but I had not tried it with '\r' since I'm on windows, it made a difference.

Let me know what other tests I can do to help resolve this. Thanks

[ssh2shell-ssh2.connect-options.zip](https://github.com/cmp-202/ssh2shell/files/464521/ssh2shell-ssh2.connect-options.zip)

**Comment by cmp-202** [comment]
Try this before responding to the last message:

 

Adding `this.emit(“msg”, “got to my keyboard-interactive handler”)` to your handler function would prove if it is being called or not. 

 

I presume the problem is the onKeyboardInteractive event handler is being defined in the host object. This wouldn’t have worked after the code was changed from the branch you had working because the host. onKeyboardInteractive() code would never get called. That was the main code difference which would definitely cause connection timeouts because your password would never get sent. 

 

New Update:

Checkout: https://github.com/cmp-202/ssh2shell#keyboard-interactive for examples of both event handler methods. Only use one, and for the code you have been using over the last few days use `SSH.on(“keyboard-interactive”,function(… ){…})`. 

Or https://github.com/cmp-202/ssh2shell/blob/master/test/keyboard-interactivetest.js for a working test example.

 

I have made the code like the previous branch that worked for you, with some extra debug and verbose messages. 

 

This is tested and works to Ubuntu server, if it doesn’t for you then you have to send me your script so I can test it.

 

To use the most recent changes either get it from git master or add ssh2shell as a dependency in package.json in the root folder of your program  then run `npm install` in the same folder to get the published version from npm. 

 

Andrew

**Comment by ajiratech2** [comment]
I must not be testing this correctly. Here is what I did:

added (without quotes) `this.emit(“msg”, “got to my keyboard-interactive handler”);`  inside "            ["keyboard-interactive"]: function(name, instructions, instructionsLang, prompts, finish){ ... }," and got an error " Unexpected token ILLEGAL" with carret below "msg".


I downloaded ssh2shell-master and pointed to that. It still fails with the following messages:

C:\Node> node sshToLantronix.js
entered mainApp
DB already exists.
Next UP - we'll connect to consoleServer
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.connect
message: Connected
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.keyboard-interactive
message: mdc-lab-slc-a.lab.wellmanage.com: Class.keyboard-interactive
message: mdc-lab-slc-a.lab.wellmanage.com: Keyboard-interactive: finish([response, array]) not called in class event handler.
message: name:
message: instructions:
message: Prompts object: [
    {
        "prompt": "Password: ",
        "echo": false
    }
]
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.error
message: mdc-lab-slc-a.lab.wellmanage.com: Class.error
message: Error: Timed out while waiting for handshake, Level: client-timeout
message: mdc-lab-slc-a.lab.wellmanage.com: Connection.close
message: Closed

Attached is a copy of my script:

[testSshToLantronix.zip](https://github.com/cmp-202/ssh2shell/files/468471/testSshToLantronix.zip)
